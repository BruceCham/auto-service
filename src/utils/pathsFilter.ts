import { SwaggerJson } from 'src/consts';
import chalk from 'chalk';
import { getExplicitModelDeps, resolveModelDeps } from './getModelDeps';

export default function pathsFilter(
  newSwagger: SwaggerJson,
  swaggerConfig: {
    exclude?: RegExp[];
    include?: RegExp[];
    autoClearModels?: boolean;
    includeModels?: RegExp[];
  }
) {
  // 支持过滤掉某些特定的规则
  const { exclude, include, includeModels, autoClearModels = true } = swaggerConfig;
  if (Array.isArray(exclude) || Array.isArray(include)) {
    const { paths, definitions = {} } = newSwagger;
    const newDefinitions: typeof definitions = {};
    newSwagger.paths = Object.keys(paths).reduce<typeof paths>((newPaths, url) => {
      const included = include?.find(reg => url.match(reg));
      const excluded = exclude?.find(reg => url.match(reg));
      // 未配置 exclude 但是配置配置了 include
      if (exclude === undefined && include !== undefined) {
        if (included) {
          newPaths[url] = paths[url];
        }
      } else {
        if (included || !excluded) {
          newPaths[url] = paths[url];
        }
      }
      if (url in newPaths && definitions) {
        // 提取 used model
        resolveModelDeps(getExplicitModelDeps(paths[url]), definitions, newDefinitions);
      }
      return newPaths;
    }, {});
    if (autoClearModels) {
      // 清理未被依赖的 model，保留被强制包含的 model
      if (newSwagger.definitions) {
        const { definitions } = newSwagger;
        newSwagger.definitions = Object.keys(definitions).reduce<typeof definitions>(
          (newDefs, model) => {
            if (includeModels && includeModels.find(reg => !!model.match(reg))) {
              newDefs[model] = definitions[model];
              // 提取强制包含的 model 的依赖
              getExplicitModelDeps(definitions[model])?.forEach(model => {
                if (!(model in newDefs)) {
                  newDefs[model] = definitions[model];
                }
              });
            }
            return newDefs;
          },
          newDefinitions
        );
      }
      const filtered = Object.keys(definitions).filter(model => !(model in newDefinitions));
      if (filtered.length) {
        console.log(chalk.yellowBright(`[IMP] Auto 将自动清理 ${filtered.join(',')} 等 models！`));
      }
    }
  }
  return newSwagger;
}
