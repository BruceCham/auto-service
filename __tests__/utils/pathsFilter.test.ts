/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
/* eslint-disable @typescript-eslint/no-unused-vars */
import pathsFilter from 'src/utils/pathsFilter';

describe('utils/pathsFilter', () => {
  const getSwagger = () =>
    ({
      basePath: '/',
      paths: {
        '/exclude/1': {
          get: {
            responses: {
              '200': {
                schema: {
                  $ref: '#/definitions/EI1'
                }
              }
            }
          }
        },
        '/exclude/2': {
          get: {
            responses: {
              '200': {
                schema: {
                  $ref: '#/definitions/E2'
                }
              }
            }
          }
        },
        '/include/1': {
          get: {
            responses: {
              '200': {
                schema: {
                  $ref: '#/definitions/EI1'
                }
              }
            }
          }
        },
        '/include/2': {
          get: {
            parameters: [
              {
                name: 'username',
                in: 'path',
                schema: {
                  $ref: '#/definitions/I2'
                }
              }
            ],
            responses: {
              '200': {
                schema: {
                  $ref: '#/definitions/I3'
                }
              }
            }
          }
        }
      },
      definitions: {
        EI1: {
          $def: '#/definitions/EI_D1'
        },
        E2: {
          $def: '#/definitions/ED2'
        },
        I2: {
          $def: '#/definitions/ID2'
        },
        I3: {
          $def: '#/definitions/ID3'
        },
        EI_D1: {
          $def: '#/definitions/EID1_D1'
        },
        ED2: {},
        ID2: {},
        ID3: {},
        EID1_D1: {}
      }
    } as Autos.SwaggerJson);
  it('pathsFilter ok', () => {
    const swagger = pathsFilter(getSwagger(), {
      exclude: [/\/exclude\//]
    });
    expect(swagger).toMatchSnapshot();
  });

  it('pathsFilter includeModels ok', () => {
    const swagger = pathsFilter(getSwagger(), {
      exclude: [/\/exclude\//],
      includeModels: [/^E2$/g]
    });
    expect(swagger).toMatchSnapshot();
  });
  it('pathsFilter keepModels ok', () => {
    const swagger = pathsFilter(getSwagger(), {
      exclude: [/\/exclude\//],
      autoClearModels: false
    });
    expect(swagger).toMatchSnapshot();
  });
});
