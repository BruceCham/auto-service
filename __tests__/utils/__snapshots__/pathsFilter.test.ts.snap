// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/pathsFilter pathsFilter includeModels ok 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "E2": Object {
      "$def": "#/definitions/ED2",
    },
    "ED2": Object {},
    "EI1": Object {
      "$def": "#/definitions/EI_D1",
    },
    "EID1_D1": Object {},
    "EI_D1": Object {
      "$def": "#/definitions/EID1_D1",
    },
    "I2": Object {
      "$def": "#/definitions/ID2",
    },
    "I3": Object {
      "$def": "#/definitions/ID3",
    },
    "ID2": Object {},
    "ID3": Object {},
  },
  "paths": Object {
    "/include/1": Object {
      "get": Object {
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/EI1",
            },
          },
        },
      },
    },
    "/include/2": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "username",
            "schema": Object {
              "$ref": "#/definitions/I2",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/I3",
            },
          },
        },
      },
    },
  },
}
`;

exports[`utils/pathsFilter pathsFilter keepModels ok 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "E2": Object {
      "$def": "#/definitions/ED2",
    },
    "ED2": Object {},
    "EI1": Object {
      "$def": "#/definitions/EI_D1",
    },
    "EID1_D1": Object {},
    "EI_D1": Object {
      "$def": "#/definitions/EID1_D1",
    },
    "I2": Object {
      "$def": "#/definitions/ID2",
    },
    "I3": Object {
      "$def": "#/definitions/ID3",
    },
    "ID2": Object {},
    "ID3": Object {},
  },
  "paths": Object {
    "/include/1": Object {
      "get": Object {
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/EI1",
            },
          },
        },
      },
    },
    "/include/2": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "username",
            "schema": Object {
              "$ref": "#/definitions/I2",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/I3",
            },
          },
        },
      },
    },
  },
}
`;

exports[`utils/pathsFilter pathsFilter ok 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "EI1": Object {
      "$def": "#/definitions/EI_D1",
    },
    "EID1_D1": Object {},
    "EI_D1": Object {
      "$def": "#/definitions/EID1_D1",
    },
    "I2": Object {
      "$def": "#/definitions/ID2",
    },
    "I3": Object {
      "$def": "#/definitions/ID3",
    },
    "ID2": Object {},
    "ID3": Object {},
  },
  "paths": Object {
    "/include/1": Object {
      "get": Object {
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/EI1",
            },
          },
        },
      },
    },
    "/include/2": Object {
      "get": Object {
        "parameters": Array [
          Object {
            "in": "path",
            "name": "username",
            "schema": Object {
              "$ref": "#/definitions/I2",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "schema": Object {
              "$ref": "#/definitions/I3",
            },
          },
        },
      },
    },
  },
}
`;
